Style guide re-write

gilesgas opened the pull request:
The aim of this work is to re-write our existing style guide to split it out into two components:

* A writing style guide.
* A Markdown syntax style guide.

Our current style guide mixes the two of these into a single guide (which we felt would benefit from being split into separate guides). Also, I felt that the writing style guide could benefit somewhat from having more explanations about language usage, along with examples to demonstrate the explanations.

gilesgas commented:
Hi @mbelton-buildkite , this is ready for you to review whenever you've got the chance.

This is pretty much my first cut - I didn't quite get as much done as I wanted to, but essentially I've done the following in this PR:

* Fleshed out more of the 'language' section and refactored some of the 'style and formatting' section. There looked like there were was some duplication throughout the guide - e.g. heading styles and formatting, so I consolidated these down into a single section about headings.
* Split out the Markdown syntax section of the 'style and formatting' section was moved out into its own guide about Markdown syntax guide.

Since I haven't had a chance to work through the rest of the existing style guide material - i.e. 'Code and filenames', 'Working with the docs site', and 'Screenshots', I've moved these to the 'Markdown syntax guide' (seemed the most appropriate spot for the time being). At least the 'filenames' and much of 'Working with the docs site' could be moved to a new 'contributors' guide in future, which could incorporate how PRs are conducted.

Couple of other points:

* I've changed 'we' and 'our' references in the styleguide content, since we say that the usage of 'we' should be avoided in the docs site. I followed suit with 'our'.
* The README file ('Updating the navigation' section) has a 'style guide' link for updating the GraphQL docs (a location within the current 'style.md' file). However, this content seems to have been removed from the current style guide before I started. Do you know where this may have been moved to?

gilesgas commented on styleguide/markdown-syntax-style.md:
```diff
@@ -1,157 +1,88 @@
+* [Working with the docs site](#working-with-the-docs-site)
+* [Screenshots](#screenshots)
+<!---
+* [GraphQL API schema](#graphql-api-schema)
```
This appears to have been moved somewhere else...

gilesgas replied:
Should this link here?
https://github.com/buildkite/docs/blob/style-guide-updates-language-and-markdown-syntax/data/graphql/schema.graphql

gilesgas replied:
As discussed, we'll leave a link to this schema in the README.

For some reason, the README also has a link on how to update the GraphQL docs (as part of its 'Updating the navigation') section, although I'm not sure if this is still a thing?
The 'Working with the docs site' section (now in the Markdown syntax style guide) has nothing on GraphQL.

gilesgas commented on styleguide/writing-style.md:
```diff
@@ -17,14 +17,14 @@ This section covers everything related to the language and formatting used in Bu
 ### English flavor
 
-Use US English. As a highly multi-national team, here is the list of the most notable [differences between American, British, and Australian English](http://web.archive.org/web/20180925084522/http://www.linktranslation.com.au/the-differences-between-american-british-and-australian-english/) English to watch out for.
+Use US English. As a highly multi-national team, here is the list of the most notable differences between [Australian, American (US), and British English](https://blog.e2language.com/australian-english-vs-american-english-vs-british-english/) English to watch out for.
```
I thought it might be more appropriate to update this external reference to one which is no longer archived (and hence may disappear at any time, a bit like deprecated code).

I searched around for a couple of alternatives and this one seemed most appropriate.

mbelton-buildkite commented on styleguide/markdown-syntax-style.md:
```diff
@@ -218,7 +149,7 @@ Use escaping to prevent this.
 ## Working with the docs site
 
-Our docs website is a custom build. This section gives some guidance on working with the setup.
+The Buildkite docs website is a custom built one. This section gives some guidance on working with the setup.
```
Nitpick: Maybe just me but I think this reads a little more smoothly:

```suggestion
The Buildkite docs is a custom-built website. This section gives some guidance on working with the setup.
```

gilesgas replied:
Thanks, legacy content I didn't entirely see through its revision.

mbelton-buildkite commented on styleguide/markdown-syntax-style.md:
```diff
@@ -1,157 +1,88 @@
 ### Code formatting
 
-We use the GitHub flavor of markdown for [formatting code](https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code).
+The Buildkite docs uses the GitHub flavor of markdown for [formatting code](https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code).
```
Oh, I see below that I'm wrong about using the GitHub flavor of Markdown. Ignore.

mbelton-buildkite commented on styleguide/markdown-syntax-style.md:
```diff
@@ -1,157 +1,88 @@
-
-No multi-sentence headings or full stops at the end of a sentence in page or sentence headings.
-Be consistent about heading levels - H1, H2, H3 - no jumping from H1 to H3 or H2 to H4.
+| Links                | Relative paths to other docs, full paths to anything else | Always full paths, check for HTTPS and that you're not using .localhost | Always full paths, check for HTTPS and that you’re not using .localhost |
```
It looks like the link information is also included in the Working with the site section

gilesgas replied:
Yeah, because I haven't been through the entire style guide content, there's some stuff I've not focussed on.

I'm thinking of having a separate section on 'Links' as part of the 'Style and formatting' section (both in the Writing and Markdown syntax style guides) on what text to link, and how to do it in Markdown, respectively.

gilesgas replied:
For the Docs column, spell out where contributors copy the URL slug from.

gilesgas replied:
Ditch table formatting for this and mention (in a note) when creating links to Twitter and Changelog from the Docs, do as instructed.

gilesgas replied:
I decided to completely refactor the existing 'Links' and 'Anchor links' sections of this page and move them higher up into the 'Style and formatting' section.

mbelton-buildkite commented on styleguide/markdown-syntax-style.md:
```diff
@@ -444,7 +378,7 @@ Line 3, column 1  | Line 3, column 2
 
 #### Responsive tables
```
The information about callouts and tables feels like it might be in the wrong section to me. I know it's technically something custom on top of Markdown, but I'm not sure if that distinction is important to our authors. I expect they're just looking how to format something to work in our site. 🤔

gilesgas replied:
Done.

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+
+For details about the Markdown syntax used to render your writing as well as the Buildkite docs' file structure, as well as details on working with the site, and using screenshots, refer to the [Markdown syntax style guide](markdown-syntax-style.md).
+
+Table of contents (main headings):
+* [Language](#language)
```
Nitpicks:

- Add a blank line before the start of a list.
- "(main headings)" feels unnecessary to me.
- I have a preference for `-` in lists, but I can change. 😅

gilesgas replied:
> Add a blank line before the start of a list.

Whoops, well spotted!

> "(main headings)" feels unnecessary to me.

Yeah, my main reasoning here was to make the distinction between this list and GitHub's auto-generated outline. I've removed it.

> I have a preference for `-` in lists, but I can change. 😅

Sorry, I thought you were referring to the lead-in to a list rather than the Markdown syntax symbol for a bullet point list item, which I think is what you meant here lol

For the lead in, it looks like the [Google style guide](https://developers.google.com/style) tend to implement colons for these, although I think we're on the same page with this one.

For the Markdown syntax symbol, I've noticed more instances of an asterisk `*` than a hyphen/minus `-`, so maybe we should choose asterisks `*`? That way, it'll be easier to update the docs at a later point in time.

Note to self—I should add this to the Markdown syntax style guide about bulleted lists.

gilesgas replied:
Michael 😢 and I agreed to stick to asterisks due to reason above.

gilesgas replied:
This is all now covered.

gilesgas replied:
Darn, I've only just discovered from the linting results that the first bullet point must be a hyphen `-` and nested can then be asterisks `*` 😲 

I'll update the guidance accordingly 🤦

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+
+### Commas
+
+Keep it clean and uncluttered. If the sentence can live without that comma - leave it out.
```
Is this right? The second sentence seems like a comma would be more appropriate?

This also feels a bit strange as guidance beside a point about using serial commas. I wonder if it should be something that emphasizes clarity not clutter, like:

> Keep it clean and uncluttered. Use commas for clarity, not decoration.

gilesgas replied:
I like that, less contradictory with the next paragraph.

I've decided to remove the distinction between the 'Commas' and 'Serial commas' sections and place all this content under a single heading 'Commas'.

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+
+### Serial commas
+
+Absolutely yes, yes, and yes. Use serial commas _when required_ - typically when listing items in a sentence.
```
Nitpick: As noted earlier, use em dashes to break up sentences.

```suggestion
Absolutely yes, yes, and yes. Use serial commas _when required_—typically when listing items in a sentence.
```

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+However, the fully active voice version (above) is preferable for instructional step-by-step content like this.
+
+Aim to use active voice in instructional step-by-step content and avoid passive voice.\
+Passive voice is useful when you want to emphasize the object, for example when describing the consequence of an activity in the previous sentence. However, stick to active voice for such sentences unless the object requires emphasis.
```
Another use case that might be worth mentioning here is using passive voice to avoid placing blame on the user.

gilesgas replied:
Oh OK, let's chat about this as I'd like to clarify this for the style guide.

gilesgas replied:
E.g. "The system crashed by the tape being inserted incorrectly" (passive and more polite) versus "You inserted the tape incorrectly leading to the system crashing".

gilesgas replied:
Added section about this.

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+
+### Product Names (and product features)
+
+Only use Title Case (initial capital letters) for the name of the product (product names), and match capital letter usage in product names as an organization would use them in these products - for example:
```
Nitpick: I'd keep it simple and start a new sentence for the last part.

```suggestion
Only use Title Case (initial capital letters) for the name of the product (product names), and match capital letter usage in product names as an organization would use them in these products. For example:
```

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+
+### Lists (bullet lists and numbered steps)
+
+Capitalize the first word; no full stops at the end if it's not a full sentence. If it's a full sentence, give it a full stop.
```
Nitpick: Semicolons are a bit of a smell to me that the sentence is more complicated than it needs to be. Consider reworking.

gilesgas replied:
OK, maybe I'll use the em dash approach instead?

gilesgas replied:
On another note, I'm changing 'full stop' to 'full stop/period'.

In the US, from my personal experience, many people haven't heard of the term 'full stop', until you tell them that it's a 'period'.

On yet another note, and I'm not sure how far we're prepared to go with this one, in US English, if there are quotation marks around a final or final set of words in a sentence, the final punctuation character is written before the final quotation mark and not after it, as is the case with British/Australian English.

For example:

* In US English, this sentence would be written "like this."

* In British/Australian English, this sentence would be written "like this".

This discrepancy results from an historical book on writing style, created around the time of typeset printing presses, when placing a period block at the end of a sentence would damage these printing presses more easily, and instead, placing the period before the final character—that is, the quotation mark—would help preserve them. I think once these presses were automated and had the print in embossed plates, this wasn't such an issue. A newer book on writing style came out after these newer presses were produced and switched the order of the final punctuation letter. However, this later style guide was only produced in Britain and its former colonies.

gilesgas replied:
Michael and I discussed - add word '/period'.

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+
+### Writing numbers
+
+Write out numbers up to 10, then use digits - '58 bugs in this script and just two hours to fix them all!'
+Long numbers use commas to separate thousands - '100,000,000.00'. When in doubt, see what [Google](https://developers.google.com/style/numbers) and [Microsoft](https://docs.microsoft.com/en-us/style-guide/numbers) say about numbers.
```
I continue my vendetta against using dashes to break up sentences:

```suggestion
Write out numbers up to 10, then use digits—'58 bugs in this script and just two hours to fix them all!'
Long numbers use commas to separate thousands—'100,000,000.00'. When in doubt, see what [Google](https://developers.google.com/style/numbers) and [Microsoft](https://docs.microsoft.com/en-us/style-guide/numbers) say about numbers.
```

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+Write out numbers up to 10, then use digits - '58 bugs in this script and just two hours to fix them all!'
+Long numbers use commas to separate thousands - '100,000,000.00'. When in doubt, see what [Google](https://developers.google.com/style/numbers) and [Microsoft](https://docs.microsoft.com/en-us/style-guide/numbers) say about numbers.
+
+Do not use numbers in page headings. Only use numbers less than 10 in section headings.
```
I wonder if we want to soften this:

```suggestion
Avoid using numbers in page headings. Only use numbers less than 10 in section headings.
```

I can see cases where we might want to do that. For example, a multi-part tutorial.

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+
+|                      | Docs                                                      | Twitter and Blog                                                        | Changelog                                                               |
+|----------------------|-----------------------------------------------------------|-------------------------------------------------------------------------|-------------------------------------------------------------------------|
+| We (as in Buildkite) | No                                                        | Yes                                                                     | Avoid if possible; use 'I' if you need to                               |
+| Exclamation marks    | No                                                        | Yes, although use with restraint, and not more than one in a row        | Yes, although use with restraint, and not more than one in a row        |
```
I agree with the sentiment, but wonder if this is too restrictive. I avoid "we" in almost every case, but do use it with "We recommend..." Similarly with exclamation marks, using one in a tutorial to celebrate success feels appropriate. What do you think about those two cases?

gilesgas replied:
Happy to make some exceptions here and there (and I'm happy with this one too), although ideally we should spell these out in the style guide.

gilesgas replied:
Michael and I agreed - simplify table to remove Twitter and changelog entries and provide generic guidance of usage of 'we' and exclamation marks (that is, avoid) unless tone of writing warrants it, such as a tutorial.

gilesgas replied:
Transformed into a new section on 'formality'.

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+| Signup/login              | When referring to a page that enables signing up or to the signup process (that is, the adjective or noun form of these terms)                       |
+| Time out/timeout          | Time out is a verb, timeout is a noun or adjective                                                            |
+| API, SSO, SAML            | Always capitalized                                                                               |
+| GitHub                    | Always capitalized, with an uppercase H in the middle                                            |
```
Some general guidance that's missing that you might consider adding is to follow brand capitalization.

gilesgas replied:
I might either remove this (because it's covered above in the **Product names (and product features)**), or repurpose this slightly to make it more general about following brand capitalisation, as I suspect you're alluding to.

gilesgas replied:
Do the point I mentioned above.

gilesgas replied:
Added "premises" to this table and removed GitHub.

mbelton-buildkite commented on styleguide/writing-style.md:
```diff
@@ -0,0 +1,183 @@
+
+Linters cannot do all of the work for you, so please pay attention to the following cases:
+
+* **Their/they’re/there, your/you’re** - [check](https://www.dictionary.com/e/their-there-theyre/ ) if you’ve got the right one for your situation!
+* **Affect/effect** - affect is a verb, effect is a noun. When you affect something, you’re impacting or changing the thing. When you have an effect, it’s the outcome or result of a change.
```
I see you hyphens! I think these work better as colons:

```suggestion
* **Their/they’re/there, your/you’re:** [Check](https://www.dictionary.com/e/their-there-theyre/ ) if you’ve got the right one for your situation!
* **Affect/effect:** Affect is a verb, effect is a noun. When you affect something, you’re impacting or changing the thing. When you have an effect, it’s the outcome or result of a change.
```

mbelton-buildkite reviewed (commented):
This is a bit of a grab bag. And I realize some go beyond the scope of what you're doing. Feel free to disregard those for now.

It might be easier to talk through a lot of these if you'd rather book in a time.

gilesgas commented on styleguide/markdown-syntax-style.md:
```diff
@@ -1,157 +1,88 @@
+> To improve the readability of the Markdown source content, ensure there is an empty line inserted both above and below the heading.
 
-> **Note:** Markdown supports two characters as its markup for itaclising text - either an underscore "\_" or a single asterisk "\*". For consistency, use single underscores "\_" immediately surrounding the text you want to italicise - e.g. `_Italicise this text_`
+Refer to [Heading capitalization in the Writing style guide](writing-style.md#heading-capitalization) for details on how to write and present headings in the Buildkite docs.
```
I suspect I may have fixed this soon after this comment of yours:
https://github.com/buildkite/docs/pull/2635/commits/a02949b92ee4e31a044bf9b45be599fbe7344b15

gilesgas commented on styleguide/markdown-syntax-style.md:
````diff
@@ -1,157 +1,88 @@
 
-Only capitalize the name of the product - for example, 'GitHub organization' and not 'GitHub Organization'.
+```
+# Page title
````
Good call, I'll update all these references; referring to `# Heading level 1` as used as the page title clears makes this relationship between Markdown much clearer.

gilesgas commented on styleguide/markdown-syntax-style.md:
```diff
@@ -1,157 +1,88 @@
+### UI elements
 
-Use 24hr time with hours and minutes, but not seconds. Include timezone. For example, 17:00 AEST
+UI element references are formatted using italics in the Buildkite docs. Markdown supports two characters as its markup for italicizing text - either an underscore "\_" or a single asterisk "\*". For consistency, use single underscores "\_" immediately surrounding the text you want to italicize - e.g. `_Italicize this text_`
```
Interesting! I like the idea of using an em dash for this purpose but wasn't aware of the lack of the 'no spacing' around it. I was a little taken aback by this but I don't have strong opinions about it, and I'm soon getting used to it :-)
Since we've decided on the Google develop documentation style guide as our main set of guidelines. I'm happy to go along with this.
I've also now included the references to these Google style guide (+ Microsoft Style Guide) at the top of the Markdown syntax style guide.

gilesgas commented on styleguide/markdown-syntax-style.md:
```diff
@@ -1,157 +1,88 @@
 
-A little [historical background](https://www.onlinegrammar.com.au/the-grammar-factor-spacing-after-end-punctuation-capitals/) on why this is even a valid question.
+A little [historical background](https://www.onlinegrammar.com.au/the-grammar-factor-spacing-after-end-punctuation-capitals/) on why this is even a valid question.\
 P.S. Remember that, ironically enough, in Markdown, line breaks demand exactly two blank spaces at the end of the line.
```
Many of these remaining comments you've made relate to existing content I haven't yet reviewed in detail.
Nevertheless, I'll address them here.

Interestingly, a final backslash ( \\ ) character, while in GitHub Markdown is treated as a line break, this is output literally in the Buildkite docs. If you want to create a line break in the Buildkite docs, you need to use the HTML tag `<br/>`. Personally, I think this is a formatting feature that could provide some value in our docs - improved readability?

I'll reword that final P.S. as it sounds like quite a tone shift lol

gilesgas replied:
Discussions with Michael:

* Ditch the `<br/>` idea - stick with Google style guide due to no other good reason for this type of formatting.
* Move old PS content into a brief section about paragraphs.

gilesgas commented on styleguide/markdown-syntax-style.md:
```diff
@@ -1,157 +1,88 @@
 
 ### Platform differences
 
+This table summarizes Markdown syntax differences across different platforms, distinguishing them from the Buildkite 'Docs'.
```
Good question, no good reason and happy to change this to "Buildkite Docs", in line with the icon/logo at the top-left of the site.

gilesgas commented on styleguide/markdown-syntax-style.md:
```diff
@@ -1,167 +1,102 @@
-A code example longer than a couple of words that isn’t part of a sentence/a multi-line code sample needs to be formatted as a code block according to the [GitHub markdown flavor](https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code).
-To add a code block, indent it using four (4) spaces or use 3 backticks (\`\`\`) before and after the code block.
+A code example longer than a couple of words that isn’t part of a sentence/a multi-line code sample needs to be formatted as a code block according to the [GitHub Markdown flavor](https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code).
+To add a code block, indent it using four (4) spaces or use three (3) backticks (\`\`\`) before and after the code block.
```
Could we omit the reference to 4 spaces and just stick with one style—3 backticks?

gilesgas replied:
Good, do so.

gilesgas commented:
Add in Headings section of Markdown syntax guide, do not use bold text for headings.

gilesgas commented:
FYI, as part of this PR, I've updated the introductory content to the README to better lay out the set-up process.

I've decided to promote the non-containerised development processes over the containerised ones since the containerised one is still a little flaky.

I also added a reference to Yarn needing to be installed too as I couldn't get the non-containerised environment up and running without it.

mbelton-buildkite reviewed (approved):
Reviewed in person :)
